File: Grind169/Week1/20-Valid-Parentheses/solution.cpp
#include <string>
#include <unordered_map>
#include <stack>

using namespace std;


class Solution {
public:
    bool isValid(string s) {
        stack<char>st;

        for (char c:s){
            if (c=='(' || c=='{' || c=='['){
                st.push(c);
            } else {
                if (st.empty() || (st.top()=='(' && c!=')')|| (st.top()=='[' && c!=']')|| (st.top()=='{' && c!='}')) {
                    return false;
                }
                else {
                    st.pop();
                }
            }
        }
        return st.empty();
        
    }
};

File: Grind169/Week1/20-Valid-Parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = [s[0]]
        mapping = {"}": "{", "]": "[", ")": "("}
        for el in s[1:]:
            if not stack:
                stack.append(el)
                continue
            if mapping.get(el) == stack[-1]:
                stack.pop()
            else:
                stack.append(el)

        return len(stack) == 0
